docs/architecture/index.adoc

```diff
@@ -0,0 +1,108 @@
+
+Brew View takes the request body and attempts to parse it into a valid Beer Garden request and save it to Mongo. These steps require passing the first-level validation check. This ensures that the request is syntatically valid and meets certain basic requirements (such as the ``status`` field being a valid value).  If it fails at this it will return a 400 status code, otherwise the request ID is passed to Bartender for processing.
+
+Bartender then pulls the completed request out of the database and *validates* it. This validation is significatly more involved than the preliminary validation done by Brew View. Here are some of the validation steps:
+
+- Ensure that a system (which the correct version) exists that can service the request
+- That system has an instance matching the instance the request is addressed to
+- That system has a command matching the request's command
+- The request's command parameters meet all the constraints placed on them by the command definition
+- There are no extra command parameters
+
+If all of these conditions are met Bartender will send the request to RabbitMQ using a routing key that ensures the request will be processed by the correct plugin.
+
+RabbitMQ will place the request in the specified plugin's request queue.
+
+Plugins maintain a consumer connection to RabbitMQ awaiting messages. When a new message is placed in their queue the plugin does several things. First, it attempts to parse the message into a valid Beer Garden request. If that's successful then it checks that the request is correctly addressed. If either of those checks fail the message is discarded.
+
+If the requests passes those checks then the plugin is able to process the request. The first step is to send an update to Beer Garden setting the status for that request to ``IN_PROGRESS``. The plugin then invokes the actual command method and captures the return value. The plugin then sends an update to Beer Garden with the results and output of the method invokation.
+
+If the 'final' update fails then the completed request is placed back on the RabbitMQ queue. This is to take advantage of RabbitMQ's message durability - if the plugin goes down at this point the request completion and output will be preserved. The request will be read from the queue and placed into a periodic retry loop. The plugin will reattempt to update the request status up to a maximum of ``max_attempts`` times, waiting an increasing amount of time between attempts (up to ``max_timeout``). Requests that fail to update before reaching ``max_attempts`` will be discarded. Note that a request in this state does not prevent processing of additional requests.
```

Comment:
Might be nice to link to the config option that lets us change this value

Reply:
The are plugin config options, and we don't really have a section on plugin configuration right now. We probably should though, I'll add that to the issues.